sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubUnboundedSink.java

```diff
@@ -0,0 +1,314 @@
+ * <li>We try to send messages in batches while also limiting send latency.
+ * <li>No stats are logged. Rather some counters are used to keep track of elements and batches.
+ * <li>Though some background threads are used by the underlying netty system all actual Pubsub
+ * calls are blocking. We rely on the underlying runner to allow multiple {@link DoFn} instances
+ * to execute concurrently and hide latency.
+ * <li>A failed work item will cause messages to be resent. Thus we rely on the Pubsub consumer
+ * to dedup messages.
+ * </ul>
+ */
+public class PubsubUnboundedSink<T> extends PTransform<PCollection<T>, PDone> {
+  private static final Logger LOG = LoggerFactory.getLogger(PubsubUnboundedSink.class);
+
+  /**
+   * Maximum number of messages per publish.
+   */
+  private static final int PUBLISH_BATCH_SIZE = 1000;
+
+  /**
+   * Maximum size of a publish batch, in bytes.
+   */
+  private static final long PUBLISH_BATCH_BYTES = 400000;
+
+  /**
+   * Longest delay between receiving a message and pushing it to Pubsub.
+   */
+  private static final Duration MAX_LATENCY = Duration.standardSeconds(2);
+
+  /**
+   * Coder for conveying outgoing messages between internal stages.
+   */
+  private static final Coder<PubsubClient.OutgoingMessage> CODER = new
+      AtomicCoder<PubsubClient.OutgoingMessage>() {
+        @Override
+        public void encode(
+            PubsubClient.OutgoingMessage value, OutputStream outStream, Context context)
+            throws CoderException, IOException {
+          ByteArrayCoder.of().encode(value.elementBytes, outStream, Context.NESTED);
+          BigEndianLongCoder.of().encode(value.timestampMsSinceEpoch, outStream, Context.NESTED);
+        }
+
+        @Override
+        public PubsubClient.OutgoingMessage decode(
+            InputStream inStream, Context context) throws CoderException, IOException {
+          byte[] elementBytes = ByteArrayCoder.of().decode(inStream, Context.NESTED);
+          long timestampMsSinceEpoch = BigEndianLongCoder.of().decode(inStream, Context.NESTED);
+          return new PubsubClient.OutgoingMessage(elementBytes, timestampMsSinceEpoch);
+        }
+      };
+
+  // ================================================================================
+  // ShardFv
+  // ================================================================================
+
+  /**
+   * Convert elements to messages and shard them.
+   */
+  private class ShardFn extends DoFn<T, KV<Integer, PubsubClient.OutgoingMessage>> {
+    private final Aggregator<Long, Long> elementCounter =
+        createAggregator("elements", new Sum.SumLongFn());
+
+    @Override
+    public void processElement(ProcessContext c) throws Exception {
+      elementCounter.addValue(1L);
+      byte[] elementBytes = CoderUtils.encodeToByteArray(elementCoder, c.element());
+      long timestampMsSinceEpoch = c.timestamp().getMillis();
+      c.output(KV.of(ThreadLocalRandom.current().nextInt(numShards),
```

Comment:
it is preferable (in terms of serialization overhead, which I understand is important for streaming) to make these `static` inner classes and then pass in constants in constructors. But up to you.